tls: deprecate NODE_TLS_REJECT_AUTHORIZED

dcposch opened the pull request:
Fixes #5258
### Included in this PR
- Deprecates the `NODE_TLS_REJECT_AUTHORIZED` environment variable
- Replaces it with `NODE_TLS_INSECURELY_ACCEPT_UNTRUSTED_CERTIFICATES`

Also thinking of adding type validation for the `rejectUnauthorized` option, so that it must be the boolean `false` to disable cert checking, rather than any falsey value, as follows:

``` js
 // require `false`, not just a "falsey" value, to forego cert validation
 assert(typeof options.rejectUnauthorized === 'boolean',
        'options.rejectUnauthorized is not boolean: ' +
        options.rejectUnauthorized);
```

Let me know if this is likely to break existing code.
### Not included
- The option is still called `rejectUnauthorized`. We could deprecate that as well and replace it with `insecurelyAcceptUntrustedCerts` or similar, but I don't know if it's worth it --- @ChALkeR found that the existing option is used in a lot of npm modules.

ChALkeR commented:
Actually, I would like to see a warning when someone is using even the new env var.
The env var could be set in a lot of places and that could be unnoticed by the user.
Don't take this a call for immediate action in amending this PR, though, I would like to know what does everyone else think.

jasnell commented:
/cc @nodejs/crypto

rvagg commented:
I think I'm +1 on always printing a warning when using the env var and in fact, if we're going to do that I don't see a good reason to deprecate `NODE_TLS_REJECT_UNAUTHORIZED` since you'll get a warning no matter what you use and `NODE_TLS_INSECURELY_ACCEPT_UNTRUSTED_CERTIFICATES` is kind of patronising.

I can see (and have personally had) valid reasons to use `rejectUnauthorized: false` but doing it app-wide with an env var is not such a great practice.

Also, has anyone bothered to dig in to the history to see _why_ an env var was introduced in the first place? That might tell an interesting story.

bnoordhuis commented:
> Also, has anyone bothered to dig in to the history to see why an env var was introduced in the first place?

https://github.com/nodejs/node-v0.x-archive/issues/3949 - it's an escape hatch (and I needed it to keep the test suite passing.)

dcposch commented:
@rvagg @ChALkeR cool
* I've removed the new env var. No more `NODE_TLS_INSECURELY_ACCEPT_UNTRUSTED_CERTIFICATES`
* It warns you when you use the old, existing env var.
* I've removed all references to the old env var from tests, so the escape hatch @bnoordhuis mentioned might not be necessary anymore.

The warning still says the old env var is deprecated. Does that look right? The idea is that in some future release, we can get rid of the env var entirely.

ChALkeR commented:
@bnoordhuis What do you think — is it still needed? Can we get rid of it?
It looks like a good idea to me to deprecate it completely.

ChALkeR commented on lib/_tls_wrap.js:
```diff
@@ -959,13 +960,25 @@ function normalizeConnectArgs(listArgs) {
   var options = args[0];
   var cb = args[1];
 
+  // these two environment variables cause node to accept untrusted certificates
```
two?

dcposch replied:
fixed

dcposch commented:
So I wanted to update the docs to say that `NODE_TLS_REJECT_UNAUTHORIZED` will now produce a deprecation warning... looks like it isn't actually documented at all!

I think that strengthens the case for removing it.

bnoordhuis commented on lib/_tls_wrap.js:
```diff
@@ -959,13 +960,25 @@ function normalizeConnectArgs(listArgs) {
   var cb = args[1];
 
+  // this environment variable causes node to accept untrusted certificates
+  // by default (that is, unless the options explicitly specify otherwise)
```
Can you capitalize and punctuate this comment and the one below?

dcposch replied:
done

bnoordhuis commented on lib/_tls_wrap.js:
```diff
@@ -959,13 +960,25 @@ function normalizeConnectArgs(listArgs) {
+      'NODE_TLS_REJECT_UNAUTHORIZED is deprecated and may make your program ' +
+      'insecure against man-in-the-middle attacks. Are you sure? If so, ' +
+      'please use the rejectUnauthorized option in your code.',
+      warned);
```
Can you indent by four spaces here?

bnoordhuis commented on test/parallel/test-https-client-get-url.js:
```diff
@@ -31,7 +29,12 @@ var server = https.createServer(options, function(req, res) {
+    path: '/foo?bar',
+    rejectUnauthorized: false
+  });
 });
```
Replacing the URL with an object changes the meaning of this test.

dcposch replied:
Yeah, I see what you mean. OTOH having the undocumented environment variable that globally disables certificate checks for this one test seems janky.

Here's the code in `https.js`:
```js  if (typeof options === 'string') {
    options = url.parse(options);
    if (!options.hostname) {
      throw new Error('Unable to determine the domain name');
    }
  } else {
    options = util._extend({}, options);
  }
```

Maybe we can leave it as `https.get(<object>)` here and test URL parsing separately? We have some tests in `test/parallel/test-http-url.parse-*.js`

I can also add a case that calls 'https.get()` on a path with no hostname, and check that it throws the right exception.

bnoordhuis replied:
> Maybe this test can pass an object to https.get(), and we can test URL parsing separately?

You might as well remove the test altogether in that case, its sole purpose is verifying that `https.get()` accepts a URL.

dcposch replied:
So it sounds like options are:
* Delete the test
* Keep the environment variable that disables certification verification
* Keep the test, but load a URL with a valid cert (for example, `https://letsencrypt.org/`)

Do the node tests run without internet access?

LMK what you prefer

Trott replied:
> Can node tests assume internet access?

Tests that run regularly do not go out to the internet (as far as I'm aware). The tests that do go out to the internet end up in `test/internet` where they are run infrequently. (Regular CI runs, for example, do not exercise those tests. It is possible that they are running during the release process or at other times. I don't know. Maybe someone who does know can chime in.)

dcposch replied:
@Trott cool, I've moved `test-https-client-get-url` to `tests/internet`.

I also made it a lot more comprehensive. It tries `https.get` on all of the different subdomains on https://badssl.com, and for each one it either expects success or failure.

Which certificates node considers valid is almost, but *not quite the same* as Chrome. Maybe @indutny can weigh in on whether this looks OK:

```js
// Make sure https.get() works with standard HTTPS settings
expectSuccess('https://badssl.com');
expectSuccess('https://dh2048.badssl.com');
expectSuccess('https://sha256.badssl.com');
expectSuccess('https://1000-sans.badssl.com');

// Unusual or bad, but valid. Accepted by both node and Chrome.
expectSuccess('https://rsa8192.badssl.com');
expectSuccess('https://very.badssl.com');
expectSuccess('https://dh1024.badssl.com');
expectSuccess('https://sha1-2017.badssl.com');

// ACCEPTED by node but REJECTED by Chrome (!)
expectSuccess('https://pinning-test.badssl.com');

// REJECTED by node but ACCEPTED by Chrome (!)
expectFailure('https://incomplete-chain.badssl.com');

// Invalid. Rejected by both node and Chrome
expectFailure('https://expired.badssl.com');
expectFailure('https://wrong.host.badssl.com');
expectFailure('https://self-signed.badssl.com');
expectFailure('https://rc4.badssl.com');
expectFailure('https://dh480.badssl.com');
expectFailure('https://subdomain.preloaded-hsts.badssl.com');
expectFailure('https://superfish.badssl.com');
expectFailure('https://edellroot.badssl.com');
expectFailure('https://dsdtestprovider.badssl.com');
```

jasnell commented:
Any updates on this? @nodejs/ctc

dcposch commented:
@Trott @bnoordhuis cool, I've moved `test-https-client-get-url` to `tests/internet`.

To try it out: `make test-internet`

I also made it more comprehensive. It now tests `https.get` on all of the different subdomains listed at https://badssl.com, and for each one it either expects success or failure.

Which certificates node considers valid is *almost, but not quite the same as Chrome*. Maybe @indutny can weigh in on whether this looks OK:

```js
// Make sure https.get() works with standard HTTPS settings
expectSuccess('https://badssl.com');
expectSuccess('https://dh2048.badssl.com');
expectSuccess('https://sha256.badssl.com');
expectSuccess('https://1000-sans.badssl.com');

// Unusual or bad, but valid. Accepted by both node and Chrome.
expectSuccess('https://rsa8192.badssl.com');
expectSuccess('https://very.badssl.com');
expectSuccess('https://dh1024.badssl.com');
expectSuccess('https://sha1-2017.badssl.com');

// ACCEPTED by node but REJECTED by Chrome (!)
expectSuccess('https://pinning-test.badssl.com');

// REJECTED by node but ACCEPTED by Chrome (!)
expectFailure('https://incomplete-chain.badssl.com');

// Invalid. Rejected by both node and Chrome
expectFailure('https://expired.badssl.com');
expectFailure('https://wrong.host.badssl.com');
expectFailure('https://self-signed.badssl.com');
expectFailure('https://rc4.badssl.com');
expectFailure('https://dh480.badssl.com');
expectFailure('https://subdomain.preloaded-hsts.badssl.com');
expectFailure('https://superfish.badssl.com');
expectFailure('https://edellroot.badssl.com');
expectFailure('https://dsdtestprovider.badssl.com');
```

ChALkeR commented:
@dcposch This opens up fine in Chromium 49:
 
> // ACCEPTED by node but REJECTED by Chrome (!)
expectSuccess('https://pinning-test.badssl.com');

dcposch commented:
Here's the error I get. Chrome 48 on Ubuntu 15.10

![image](https://cloud.githubusercontent.com/assets/169280/13949423/2fe895c0-efe2-11e5-82fa-09d21482d9ef.png)

![image](https://cloud.githubusercontent.com/assets/169280/13949430/3eacd95e-efe2-11e5-8781-a974cbf7b45a.png)

ChALkeR commented:
So, what's the status of this?

+1 for deprecating, but new tests here don't look good to me — they introduce dependencies on third-party websites at test time, also the comments at those tests (claims about Chrome behaviour) don't correspond to the actual behaviour that I see.

ChALkeR commented:
@bnoordhuis in https://github.com/nodejs/node-v0.x-archive/issues/3949#issuecomment-8163351 that was:
> To be investigated: test/simple/test-https-pfx.js fails when you set rejectUnauthorized explicitly. The expected validation error changes from UNABLE_TO_GET_ISSUER_CERT to DEPTH_ZERO_SELF_SIGNED_CERT.

I assume that it's not required now.

jasnell commented:
Closing given the lack of forward progress on this
